proposals/052-cryptography-leg-1.md

```diff
@@ -0,0 +1,393 @@
+
+The others were not selected for several reasons:
+
+- libgcrypt is C, and so the Haskell FFI can bind directly to it, but it has a GNU LGPL license, and provides less functionality than Botan
+- Crypto++ is C++, and has a Boost license (similar to BSD), but it does not export a C API, and provides less functionality than Botan
+- libsodium / NaCL does not provide the necessary functionality, and there are several Haskell libsodium bindings already (see Appendix C - NaCl / libsodium)
+- NSS has a copyleft license
+
+Bindings to Botan would solve the significant problem of dealing with common cryptography by providing much of the necessary 'cryptographic kitchen sink' via bindings to a suitably performant, suitably licensed, open-source library. The Haskell community would not be required to maintain the Botan cryptography library, only bindings to it, and this can be accomplished readily with some effort which has already begun. Furthermore, the Botan library supports several post-quantum cryptography algorithms, which is highly desirable.
+
+Furthermore, bindings to a dedicated cryptography library may yield a performance benefit, as one would expect a popular C++ crypto library such as Botan to be highly performant. Testing in GHCi with `:set +s` for a requested example of `bcrypt` yielded the following results:
+
+```
+-- Botan
+ghci> bcryptGenerateIO "Fee fi fo fum!" r 16
+"$2a$16$j3Obv.HNT8zQ9OXFXEJeQeq3nT5WmUecPYDF0tc3grNsX9N0uDSCm\NUL"
+(3.73 secs, 996,152 bytes)
+
+-- cryptonite
+ghci> hashPassword 16 ("Fee fi fo fum!" :: ByteString) :: IO ByteString
+"$2b$16$nmHHDe/wnfSl85dIFtqjMuAHxqAmL.JQ6knp8e6kf.X1V9/LCZxUy"
+(8.84 secs, 2,138,266,840 bytes)
+```
+
+It was quick and dirty testing in GHCi so your mileage may vary, but the results indicate that bindings to Botan may be significantly faster and consumes less memory than `crypton/ite`, if other modules / functions are similarly performant. Further implentation, testing, and benchmarking will be necessary to confirm this.
+
+# Prior Art and Related Efforts
+
+We can group Haskell cryptography efforts into several categories, which are not necessarily distinct:
+
+- Libraries for specific algorithms
+- Cryptographic "kitchen sinks"
+- Bindings to foreign libraries
+- Efforts by [Thomas DuBuisson](https://hackage.haskell.org/user/ThomasDuBuisson)
+- Efforts by [Vincent Hanquez](https://hackage.haskell.org/user/VincentHanquez)
+
+The wiki lists [several libraries](https://wiki.haskell.org/Applications_and_libraries/Cryptography) and Hackage lists [many more](https://hackage.haskell.org/packages/#cat:Cryptography) according to the category.
+
+> NOTE: Cryptocurrency libraries and libraries not on Hackage are excluded from consideration.
+
+Specific libraries worth noting are:
+
+- Predecessors to `cryptonite` - all deprecated
+    - [cryptohash](https://hackage.haskell.org/package/cryptohash)
```

Comment:
There is also a family of `cryptohash-*` packages, all kept on life support.

Reply:
Yes, I'm not quite sure what their exact relationship is to `cryptohash` (eg, do they predate it or the other way around) but they are both by the same author and presumably share a code base.

Reply:
I looked into it and it appears that those are stripped-down, algorithm-specific versions of `cryptohash` to reduce the footprint. I've added brief mention of this.